### PR TITLE
Add Mobilint ASR beam cache tracing

### PR DESCRIPTION
**Pull request by beomsu-mobilint**
## Summary

- Integrate `MobilintBeamCache` with Whisper and Qwen ASR beam-search generation paths.
- Add token-history based beam cache replay/reorder handling for Mobilint NPU ASR decoders.
- Propagate encoder/decoder core-mode benchmark kwargs for Whisper ASR benchmark runs.
- Add focused tests for BeamCache behavior, ASR CLI core-mode propagation, and Whisper/Qwen ASR beam-search paths.
- Add `scripts/debug_whisper_beam_trace.py` for comparing GPU/NPU Whisper beam-search traces.

## GPU/NPU beam-search benchmark summary

Existing ASR benchmark results under `benchmark/transformers/results/{clean,other}` show a backend-specific beam-search trend:

- GPU `openai/whisper-*`: increasing `num_beams` from 1 to 5 reduces or preserves WER/CER.
- NPU `mobilint/whisper-*`: increasing `num_beams` from 1 to 5 can increase WER/CER.
- Qwen3-ASR does not show the same Whisper-specific degradation pattern, which narrows the issue to the Whisper NPU beam-search path or model/logit behavior rather than the benchmark harness globally.

### WER comparison

Lower is better. GPU Whisper consistently improves from beam1 to beam5, while NPU Whisper often regresses, especially on `other`.

| Dataset | Model | Beam 1 WER (%) | Beam 5 WER (%) | Direction |
| --- | --- | ---: | ---: | --- |
| clean | `openai/whisper-small` | 4.17 | 3.95 | improves |
| clean | `openai/whisper-medium` | 3.62 | 3.42 | improves |
| clean | `openai/whisper-large-v3-turbo` | 2.96 | 2.85 | improves |
| clean | `mobilint/whisper-small-single` | 6.17 | 6.07 | improves slightly |
| clean | `mobilint/whisper-medium-single` | 3.85 | 4.25 | regresses |
| clean | `mobilint/whisper-large-v3-turbo-single` | 13.87 | 13.97 | regresses slightly |
| other | `openai/whisper-small` | 8.54 | 8.00 | improves |
| other | `openai/whisper-medium` | 6.77 | 6.47 | improves |
| other | `openai/whisper-large-v3-turbo` | 5.12 | 5.00 | improves |
| other | `mobilint/whisper-small-single` | 7.95 | 8.67 | regresses |
| other | `mobilint/whisper-medium-single` | 6.27 | 6.51 | regresses |
| other | `mobilint/whisper-large-v3-turbo-single` | 13.37 | 15.78 | regresses |

wer

### CER comparison

Lower is better. CER follows the same pattern as WER: GPU beam5 improves, while NPU Whisper beam5 can increase character errors.

| Dataset | Model | Beam 1 CER (%) | Beam 5 CER (%) | Direction |
| --- | --- | ---: | ---: | --- |
| clean | `openai/whisper-small` | 1.66 | 1.50 | improves |
| clean | `openai/whisper-medium` | 1.48 | 1.31 | improves |
| clean | `openai/whisper-large-v3-turbo` | 1.08 | 1.00 | improves |
| clean | `mobilint/whisper-small-single` | 2.61 | 2.57 | improves slightly |
| clean | `mobilint/whisper-medium-single` | 1.62 | 1.72 | regresses |
| clean | `mobilint/whisper-large-v3-turbo-single` | 5.25 | 6.61 | regresses |
| other | `openai/whisper-small` | 3.95 | 3.65 | improves |
| other | `openai/whisper-medium` | 3.11 | 2.90 | improves |
| other | `openai/whisper-large-v3-turbo` | 1.99 | 1.95 | improves |
| other | `mobilint/whisper-small-single` | 3.56 | 4.13 | regresses |
| other | `mobilint/whisper-medium-single` | 2.17 | 2.27 | regresses |
| other | `mobilint/whisper-large-v3-turbo-single` | 5.03 | 5.59 | regresses |

cer

### RTF comparison

Lower is better. Beam5 increases decoding cost on both GPU and NPU, but NPU beam5 has a much larger RTF increase for the measured Whisper variants.

| Dataset | Model | Beam 1 RTF | Beam 5 RTF | Notes |
| --- | --- | ---: | ---: | --- |
| clean | `openai/whisper-small` | 0.0376 | 0.0526 | GPU cost increases modestly |
| clean | `openai/whisper-medium` | 0.0681 | 0.1010 | GPU cost increases modestly |
| clean | `openai/whisper-large-v3-turbo` | 0.0255 | 0.0332 | GPU cost increases modestly |
| clean | `mobilint/whisper-small-single` | 0.0715 | 0.4241 | NPU beam5 much slower |
| clean | `mobilint/whisper-medium-single` | 0.3522 | 1.1223 | NPU beam5 much slower |
| clean | `mobilint/whisper-large-v3-turbo-single` | 0.4007 | 0.6144 | NPU beam5 slower |
| other | `openai/whisper-small` | 0.0401 | 0.0548 | GPU cost increases modestly |
| other | `openai/whisper-medium` | 0.0711 | 0.1079 | GPU cost increases modestly |
| other | `openai/whisper-large-v3-turbo` | 0.0271 | 0.0352 | GPU cost increases modestly |
| other | `mobilint/whisper-small-single` | 0.1544 | 0.4474 | NPU beam5 slower |
| other | `mobilint/whisper-medium-single` | 0.3963 | 1.1791 | NPU beam5 much slower |
| other | `mobilint/whisper-large-v3-turbo-single` | 0.4783 | 0.7044 | NPU beam5 slower |

rtf

## Whisper beam trace experiment

I added and used `scripts/debug_whisper_beam_trace.py` to compare one fixed LibriSpeech sample with GPU/NPU Whisper-small.

Configuration:

- Dataset: `openslr/librispeech_asr`, config `clean`, split `test`, shuffled with seed `0`.
- Sample id: `1284-1180-0021`.
- Model pair: `openai/whisper-small` vs `mobilint/whisper-small`.
- Generation: `num_beams=1/5`, `max_new_tokens=32`, task `transcribe`, language `en`.

Observed output:

- GPU beam1/beam5: `... catching mice, and may prove of some use`.
- NPU beam1: `... catching mice, and may prove of some use`.
- NPU beam5: `... catching mice and may prove of some use to`.

Key trace observation:

- Around the token after `mice`, NPU beam5 keeps both comma token `11` and `and` token `293` as near-tied candidates.
- The NPU logits show effectively tied top candidates in this region, e.g. token `293` and token `11` both appearing with nearly identical scores.
- Forcing full replay with `MBLT_WHISPER_BEAM_FORCE_FULL_REPLAY=1` did not change the NPU beam5 output.
- Passing `use_cache=False` also did not change the NPU beam5 output.

Conclusion from this experiment:

- The observed Whisper beam5 degradation is unlikely to be caused directly by MobilintBeamCache reorder/suffix replay corruption.
- The issue looks more consistent with beam search selecting a different near-tie punctuation path under NPU logits/quantization, especially around punctuation candidates.
- Further investigation should focus on logits parity, beam scoring tie-breaking, and Whisper punctuation/logits-processor behavior rather than cache memory replay alone.

## Validation

- `uv run ruff check mblt_model_zoo/hf_transformers/utils/cache_utils.py mblt_model_zoo/hf_transformers/models/whisper/modeling_whisper.py scripts/debug_whisper_beam_trace.py`
- Manual GPU/NPU trace runs with `scripts/debug_whisper_beam_trace.py` for beam1/beam5 Whisper-small sample comparison.

### CODE REVIEW

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review

**Comment by beomsu-mobilint** [comment]
@mobilint-review
